sycl/include/sycl/types.hpp

```diff
@@ -2420,6 +2463,14 @@ struct is_device_copyable<
                                          !std::is_trivially_copyable<T>::value>>
     : std::true_type {};
 
+// array is device copyable if element type is device copyable and it is also
+// not trivially copyable (if the element type is trivially copyable, the array
+// is device copyable by default).
+template <typename T, std::size_t N>
+struct is_device_copyable<
+    T[N], std::enable_if_t<!std::is_trivially_copyable<T>::value>>
+    : is_device_copyable<T> {};
```

Comment:
I can't find this requirement in the spec: it says that any type which is trivially copyable is device copyable, but it doesn't require us to support plain arrays of non-trivially-copyable types.

Moreover, marking such types as device copyable will likely lead to undefined behavior during kernel execution, because value may not be properly copied from device to host or vice versa. Therefore, we may be confusing users here. Am I missing something here?

Reply:
It says `legal`, not `device_copyable`. 

> plain arrays of non-trivially-copyable but device-copyable types.

That is indeed a `legal` type, i.e. it can be passed to a kernel, but I don't see that such type should be implicitly marked as `device_copyable` in [3.13.1. Device copyable](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable).

Let's maybe tag @gmlueck here, because perhaps we should tweak the spec to mark plain arrays of device copyable types as device copyable.

Reply:
I think [section 3.13.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable) already contains wording saying that an array of `T` is device copyable if `T` is device copyable:

> When the application explicitly declares a class type to be device copyable, arrays of that type and cv-qualified versions of that type are also device copyable

Reply:
> I think [section 3.13.1](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable) already contains wording saying that an array of `T` is device copyable if `T` is device copyable:
> 
> > When the application explicitly declares a class type to be device copyable, arrays of that type and cv-qualified versions of that type are also device copyable

I suppose that we have to have the specialization in question in order to support arrays of user-defined trivially copyable types. But strictly speaking, I don't see a requirement for the implementation to **implicitly** declare an array of trivially copyable types as trivially copyable types. The phrase above only covers explicitly-declared device copyable types, which may not be even supported by an implementation (`SYCL_DEVICE_COPYABLE` is not set)

Reply:
As I understand it, the type of the array is *explicitly* declared as device copyable, but the array is *implicitly* device copyable because of the rule mentioned in this discussion.

Reply:
I tried to clarify the spec in KhronosGroup/SYCL-Docs#357.  Does this address your question?

Reply:
Thanks, @gmlueck. Yes, that addresses my question.

@maarquitos14, I will mark this as resolved, because I suppose we all agree that there are no changes required, i.e. the specialization you are adding is indeed necessary for being conformant with the spec.